cmd/cstor-pool-mgmt/controller/backup-controller/handler.go

```diff
@@ -0,0 +1,237 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backupcontroller
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/golang/glog"
+	"github.com/openebs/maya/cmd/cstor-pool-mgmt/controller/common"
+	"github.com/openebs/maya/cmd/cstor-pool-mgmt/volumereplica"
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the CStorReplicaUpdated resource
+// with the current status of the resource.
+func (c *BackupController) syncHandler(key string, operation common.QueueOperation) error {
+	glog.Infof("Sync handler called for key:%s with op:%s", key, operation)
+	bkp, err := c.getBackupCStorResource(key)
+	if err != nil {
+		return err
+	}
+	if bkp == nil {
+		return fmt.Errorf("cannot retrieve BackupCStor %q", key)
+	}
+	if IsDoneStatus(bkp) || IsFailedStatus(bkp) {
+		return nil
+	}
+
+	status, err := c.eventHandler(operation, bkp)
+	if status == "" {
+		return nil
+	}
+
+	if err != nil {
+		glog.Errorf(err.Error())
+		bkp.Status = apis.BKPCStorStatusFailed
+	} else {
+		bkp.Status = apis.BackupCStorStatus(status)
+	}
+
+	nbkp, err := c.clientset.OpenebsV1alpha1().BackupCStors(bkp.Namespace).Get(bkp.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	nbkp.Status = bkp.Status
+
+	_, err = c.clientset.OpenebsV1alpha1().BackupCStors(nbkp.Namespace).Update(nbkp)
+	if err != nil {
+		return err
+	}
+
+	glog.Infof("Completed operation:%v for backup:%v, status:%v", operation, nbkp.Name, nbkp.Status)
+	return nil
+}
+
+// eventHandler will execute a function according to a given operation
+func (c *BackupController) eventHandler(operation common.QueueOperation, bkp *apis.BackupCStor) (string, error) {
+	switch operation {
+	case common.QOpAdd:
+		return c.addEventHandler(bkp)
+	case common.QOpDestroy:
+		/* TODO: Handle backup destroy event
+		 */
+		return "", nil
+	case common.QOpSync:
+		return c.syncEventHandler(bkp)
+	case common.QOpModify:
+		return "", nil
+	}
+	return string(apis.BKPCStorStatusInvalid), nil
+}
+
+// addEventHandler will change the state of backup to Init state.
+func (c *BackupController) addEventHandler(bkp *apis.BackupCStor) (string, error) {
```

Comment:
can we explain in detail in code comments what this logic tries to solve!!!

Reply:
ok